File: smart-ground-node/src/main/java/ch/jp/shooting/node/pairing/PairingCodec.java
package ch.jp.shooting.node.pairing;

import ch.jp.shooting.node.crypto.AesGcm;
import ch.jp.shooting.node.crypto.Hkdf;
import ch.jp.shooting.node.frame.FrameHeader;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.GeneralSecurityException;
import java.util.Arrays;

/**
 * Baut und parst die Pairing-Frames DISCOVER/OFFER/CONFIRM unter K_Box
 * (ADR-003, docs/superpowers/specs/2026-07-09-espnow-protocol-contracts-design.md Abschnitt 3).
 */
public final class PairingCodec {

    private static final int MIC_LENGTH = 16;
    private static final byte[] SESSION_KEY_INFO = "smart-ground-espnow-session".getBytes(StandardCharsets.UTF_8);
    private static final int SESSION_KEY_LENGTH = 32;

    private PairingCodec() {
    }

    public static byte[] buildDiscover(FrameHeader header, byte[] boxUuid, byte[] nonceB, byte[] kBox) {
        byte[] headerBytes = header.encode();
        byte[] body = concat(boxUuid, nonceB);
        byte[] mic = mic(kBox, concat(headerBytes, body));
        return concat(headerBytes, body, mic);
    }

    public static boolean verifyDiscover(byte[] frame, byte[] kBox) {
        byte[] headerBytes = Arrays.copyOfRange(frame, 0, FrameHeader.SIZE);
        byte[] body = Arrays.copyOfRange(frame, FrameHeader.SIZE, frame.length - MIC_LENGTH);
        byte[] mic = Arrays.copyOfRange(frame, frame.length - MIC_LENGTH, frame.length);
        byte[] expected = mic(kBox, concat(headerBytes, body));
        return Arrays.equals(mic, expected);
    }

    public static byte[] boxUuidOf(byte[] discoverFrame) {
        return Arrays.copyOfRange(discoverFrame, FrameHeader.SIZE, FrameHeader.SIZE + 16);
    }

    public static byte[] nonceBOf(byte[] discoverFrame) {
        return Arrays.copyOfRange(discoverFrame, FrameHeader.SIZE + 16, FrameHeader.SIZE + 24);
    }

    public static byte[] buildOffer(FrameHeader header, int radioId, int channel, byte[] nonceN, byte[] nonceB,
                                     byte[] kBox) {
        byte[] headerBytes = header.encode();
        byte[] gcmNonce = concat(new byte[4], nonceN);
        byte[] ciphertextAndTag = AesGcm.encrypt(kBox, gcmNonce, headerBytes, nonceB);
        byte[] body = concat(new byte[]{(byte) radioId, (byte) channel}, nonceN, ciphertextAndTag);
        return concat(headerBytes, body);
    }

    public static byte[] nonceBFromOffer(byte[] offerFrame, byte[] kBox) {
        byte[] headerBytes = Arrays.copyOfRange(offerFrame, 0, FrameHeader.SIZE);
        byte[] nonceN = Arrays.copyOfRange(offerFrame, FrameHeader.SIZE + 2, FrameHeader.SIZE + 10);
        byte[] ciphertextAndTag = Arrays.copyOfRange(offerFrame, FrameHeader.SIZE + 10, offerFrame.length);
        byte[] gcmNonce = concat(new byte[4], nonceN);
        try {
            return AesGcm.decrypt(kBox, gcmNonce, headerBytes, ciphertextAndTag);
        } catch (GeneralSecurityException e) {
            throw new IllegalArgumentException("PAIR_OFFER: Entschluesselung fehlgeschlagen", e);
        }
    }

    public static int radioIdOf(byte[] offerFrame) {
        return offerFrame[FrameHeader.SIZE] & 0xFF;
    }

    public static int channelOf(byte[] offerFrame) {
        return offerFrame[FrameHeader.SIZE + 1] & 0xFF;
    }

    public static byte[] nonceNOfOffer(byte[] offerFrame) {
        return Arrays.copyOfRange(offerFrame, FrameHeader.SIZE + 2, FrameHeader.SIZE + 10);
    }

    public static byte[] buildConfirm(FrameHeader header, byte[] nonceN, byte[] kBox) {
        byte[] headerBytes = header.encode();
        byte[] mic = mic(kBox, concat(headerBytes, nonceN));
        return concat(headerBytes, nonceN, mic);
    }

    public static boolean verifyConfirm(byte[] frame, byte[] kBox) {
        byte[] headerBytes = Arrays.copyOfRange(frame, 0, FrameHeader.SIZE);
        byte[] nonceN = Arrays.copyOfRange(frame, FrameHeader.SIZE, frame.length - MIC_LENGTH);
        byte[] mic = Arrays.copyOfRange(frame, frame.length - MIC_LENGTH, frame.length);
        byte[] expected = mic(kBox, concat(headerBytes, nonceN));
        return Arrays.equals(mic, expected);
    }

    public static byte[] nonceNOfConfirm(byte[] confirmFrame) {
        return Arrays.copyOfRange(confirmFrame, FrameHeader.SIZE, confirmFrame.length - MIC_LENGTH);
    }

    public static byte[] deriveSessionKey(byte[] kBox, byte[] nonceB, byte[] nonceN) {
        byte[] salt = concat(nonceB, nonceN);
        byte[] prk = Hkdf.extract(salt, kBox);
        return Hkdf.expand(prk, SESSION_KEY_INFO, SESSION_KEY_LENGTH);
    }

    private static byte[] mic(byte[] key, byte[] message) {
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(new SecretKeySpec(key, "HmacSHA256"));
            byte[] full = mac.doFinal(message);
            return Arrays.copyOf(full, MIC_LENGTH);
        } catch (GeneralSecurityException e) {
            throw new IllegalStateException("HMAC-SHA256 nicht verfuegbar", e);
        }
    }

    private static byte[] concat(byte[]... parts) {
        int total = 0;
        for (byte[] p : parts) {
            total += p.length;
        }
        byte[] out = new byte[total];
        int pos = 0;
        for (byte[] p : parts) {
            System.arraycopy(p, 0, out, pos, p.length);
            pos += p.length;
        }
        return out;
    }
}

File: smart-box/pairing_codec.py
"""Baut und parst die Pairing-Frames DISCOVER/OFFER/CONFIRM unter K_Box.

ADR-003, docs/superpowers/specs/2026-07-09-espnow-protocol-contracts-design.md Abschnitt 3.
"""
import espnow_crypto
from frame_envelope import HEADER_SIZE, TYPE_PAIR_CONFIRM, TYPE_PAIR_DISCOVER, TYPE_PAIR_OFFER, pack_header

MIC_LENGTH = 16
SESSION_KEY_INFO = b"smart-ground-espnow-session"
SESSION_KEY_LENGTH = 32


def _mic(key, message):
    return espnow_crypto.hmac_sha256(key, message)[:MIC_LENGTH]


def build_discover(dest_mac, src_mac, frame_id, ttl, box_uuid, nonce_b, k_box):
    header = pack_header(dest_mac, src_mac, frame_id, ttl, TYPE_PAIR_DISCOVER)
    body = box_uuid + nonce_b
    mic = _mic(k_box, header + body)
    return header + body + mic


def verify_discover(frame, k_box):
    header = frame[:HEADER_SIZE]
    body = frame[HEADER_SIZE:-MIC_LENGTH]
    mic = frame[-MIC_LENGTH:]
    expected = _mic(k_box, header + body)
    return mic == expected


def box_uuid_of(discover_frame):
    return discover_frame[HEADER_SIZE:HEADER_SIZE + 16]


def nonce_b_of(discover_frame):
    return discover_frame[HEADER_SIZE + 16:HEADER_SIZE + 24]


def build_offer(dest_mac, src_mac, frame_id, ttl, radio_id, channel, nonce_n, nonce_b, k_box):
    header = pack_header(dest_mac, src_mac, frame_id, ttl, TYPE_PAIR_OFFER)
    gcm_nonce = bytes(4) + nonce_n
    ciphertext_and_tag = espnow_crypto.aes256_gcm_encrypt(k_box, gcm_nonce, header, nonce_b)
    body = bytes([radio_id, channel]) + nonce_n + ciphertext_and_tag
    return header + body


def nonce_b_from_offer(offer_frame, k_box):
    header = offer_frame[:HEADER_SIZE]
    nonce_n = offer_frame[HEADER_SIZE + 2:HEADER_SIZE + 10]
    ciphertext_and_tag = offer_frame[HEADER_SIZE + 10:]
    gcm_nonce = bytes(4) + nonce_n
    return espnow_crypto.aes256_gcm_decrypt(k_box, gcm_nonce, header, ciphertext_and_tag)


def radio_id_of(offer_frame):
    return offer_frame[HEADER_SIZE]


def channel_of(offer_frame):
    return offer_frame[HEADER_SIZE + 1]


def nonce_n_of_offer(offer_frame):
    return offer_frame[HEADER_SIZE + 2:HEADER_SIZE + 10]


def build_confirm(dest_mac, src_mac, frame_id, ttl, nonce_n, k_box):
    header = pack_header(dest_mac, src_mac, frame_id, ttl, TYPE_PAIR_CONFIRM)
    mic = _mic(k_box, header + nonce_n)
    return header + nonce_n + mic


def verify_confirm(frame, k_box):
    header = frame[:HEADER_SIZE]
    nonce_n = frame[HEADER_SIZE:-MIC_LENGTH]
    mic = frame[-MIC_LENGTH:]
    expected = _mic(k_box, header + nonce_n)
    return mic == expected


def nonce_n_of_confirm(confirm_frame):
    return confirm_frame[HEADER_SIZE:-MIC_LENGTH]


def derive_session_key(k_box, nonce_b, nonce_n):
    salt = nonce_b + nonce_n
    prk = espnow_crypto.hkdf_extract(salt, k_box)
    return espnow_crypto.hkdf_expand(prk, SESSION_KEY_INFO, SESSION_KEY_LENGTH)
